### PR TITLE
Adds support for client-server multiplayer mode in the Editor with CTRL-G

### PR DESCRIPTION
**Pull request by AMZN-Olex**
## What does this PR do?

- Adds a notification when the Editor spawns the root prefab
- Adds support for client-server multiplayer in the Editor's game mode
- Use `editor_clientserver`= true to enable client-server mode in the Editor
- Adds support for RPC locally working for the following modes (this fixes https://github.com/o3de/o3de/issues/11283 )
-- Authority->Client
-- Authority->Autonomous
-- Autonomous->Authority

## How was this PR tested?

In the Editor with a custom client-server level to test all of the above features. https://github.com/o3de/o3de-multiplayersample/pull/204
Also tested that regular dedicated mode still works with `editor_clientserver`= false

image

Added new unit tests that exercise new code.
image

```
OKAY Library loaded: C:/git/o3de/build/bin/debug/Multiplayer.Tests.dll
OKAY Symbol found: AzRunUnitTests
Note: Google Test filter = *LocalRpc*
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from LocalRpcTests
[ RUN      ] LocalRpcTests.LocalRpcServerToAuthority
[       OK ] LocalRpcTests.LocalRpcServerToAuthority (303 ms)
[ RUN      ] LocalRpcTests.LocalRpcAuthorityToClients
[       OK ] LocalRpcTests.LocalRpcAuthorityToClients (88 ms)
[ RUN      ] LocalRpcTests.LocalRpcAuthorityToAutonomous
[       OK ] LocalRpcTests.LocalRpcAuthorityToAutonomous (93 ms)
[ RUN      ] LocalRpcTests.LocalRpcAutonomousToAuthority
[       OK ] LocalRpcTests.LocalRpcAutonomousToAuthority (89 ms)
[----------] 4 tests from LocalRpcTests (575 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (577 ms total)
[  PASSED  ] 4 tests.
```

### CODE REVIEW

File: Gems/Multiplayer/Code/Source/Components/NetBindComponent.cpp [hunk 1]
```diff
@@ -800,6 +811,18 @@ namespace Multiplayer
...
+    {
+        message.SetRpcDeliveryType(RpcDeliveryType::AutonomousToAuthority);
+        GetNetworkEntityManager()->HandleLocalRpcMessage(message);
+    }
+
+    void NetBindComponent::HandleLocalAuthorityToClientRpcMessage(NetworkEntityRpcMessage& message)
+    {
+        message.SetRpcDeliveryType(RpcDeliveryType::AuthorityToClient);
+        GetNetworkEntityManager()->HandleLocalRpcMessage(message);
+    }
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Source/Components/NetBindComponent.cpp:824]: Unit tests exercising these would be good.

**Comment by AMZN-puvvadar** [comment]
This looks good to me. Are there any performance differences between this mode and the original worth highlighting? Is there an accompanying docs change for the new cvar?

File: Gems/Multiplayer/Code/Source/Components/NetBindComponent.cpp [hunk 2]
```diff
@@ -177,7 +179,9 @@ namespace Multiplayer
             // The component hasn't been pre-setup with NetworkEntityManager yet. Setup now.
             const AZ::Name netSpawnableName = AZ::Interface<INetworkSpawnableLibrary>::Get()->GetSpawnableNameFromAssetId(m_prefabAssetId);
 
-            AZ_Assert(!netSpawnableName.IsEmpty(),
+            // In client-server the level asset is a temporary Root.network.spawnable and isn't registered in time.
+            // But this is not needed in client-server mode.
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Source/Components/NetBindComponent.cpp:183]:
```suggestion
            // In client-server the level asset is a temporary Root.network.spawnable and is not expected to be registered in time.
```

This isn't reading clearly to me. Maybe this suggestion is what you're going for?

File: Gems/Multiplayer/Code/Source/Editor/MultiplayerEditorSystemComponent.cpp [hunk 3]
```diff
@@ -500,6 +502,14 @@ namespace Multiplayer
             return;
         }
 
+        if (editor_clientserver)
+        {
+            // Start hosting as a client-server
+            const bool isDedicated = false;
+            AZ::Interface<IMultiplayer>::Get()->StartHosting(33450, isDedicated);
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Source/Editor/MultiplayerEditorSystemComponent.cpp:509]: Port number should be a CVAR.

File: Gems/Multiplayer/Code/Source/Editor/MultiplayerEditorSystemComponent.cpp [hunk 4]
```diff
@@ -610,4 +625,16 @@ namespace Multiplayer
             AZ::ComponentTypeList{ azrtti_typeid<NetBindComponent>(), azrtti_typeid<NetworkTransformComponent>() }
         );
     }
+    void MultiplayerEditorSystemComponent::OnStopPlayInEditorBegin()
+    {
+        if (!editor_clientserver)
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Source/Editor/MultiplayerEditorSystemComponent.cpp:630]: You might want to check AgentType here vs the CVAR in the unlikely but possible case of someone editing editor_clientserver during play mode.

File: Gems/Multiplayer/Code/Source/Components/NetBindComponent.cpp (hunk 2 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/Components/NetBindComponent.cpp:183]: Yeah, that is good.

File: Gems/Multiplayer/Code/Source/Editor/MultiplayerEditorSystemComponent.cpp (hunk 4 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/Editor/MultiplayerEditorSystemComponent.cpp]: Good idea, added a check.

File: Gems/Multiplayer/Code/Source/Components/NetBindComponent.cpp (hunk 1 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/Components/NetBindComponent.cpp:823]: Working on it now.

Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/Components/NetBindComponent.cpp]: Done.

**Comment by AMZN-Olex** [comment]
> This looks good to me. Are there any performance differences between this mode and the original worth highlighting? Is there an accompanying docs change for the new cvar?

@AMZN-puvvadar - good questions! The doc change is here: https://github.com/o3de/o3de.org/pull/2040

TLDR: the client-server mode enters game mode immediately. So it is great for quick iterative work.

**Review by AMZN-puvvadar — APPROVED** [review]
It'd be good to discuss adding a button to toggle this behavior with sig-ux in the future. Changes look reasonable to me otherwise. I'd also get @kberg-amzn 's eyes on the local server RPC Message portion of the change specifically.

File: Gems/Multiplayer/Code/Source/Editor/MultiplayerEditorSystemComponent.cpp [hunk 5]
```diff
@@ -45,6 +45,8 @@ namespace Multiplayer
 {
     using namespace AzNetworking;
 
+    AZ_CVAR(bool, editor_clientserver, false, nullptr, AZ::ConsoleFunctorFlags::DontReplicate,
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Source/Editor/MultiplayerEditorSystemComponent.cpp:48]: Noticed this in the docs change but should this be editorsv_clientserver for consistency?

Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/Editor/MultiplayerEditorSystemComponent.cpp]: Done.

**Review by AMZN-puvvadar — APPROVED** [review]

**Review by AMZN-Gene — APPROVED** [review]
